homeassistant/components/rfxtrx/siren.py

```diff
@@ -0,0 +1,228 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant, callback
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_call_later
+
+from . import (
+    DEFAULT_SIGNAL_REPETITIONS,
+    DeviceTuple,
+    RfxtrxCommandEntity,
+    async_setup_platform_entry,
+)
+from .const import CONF_SIGNAL_REPETITIONS
+
+SUPPORT_RFXTRX = SUPPORT_TURN_ON | SUPPORT_TONES
+
+SECURITY_PANIC_ON = "Panic"
+SECURITY_PANIC_OFF = "End Panic"
+SECURITY_PANIC_ALL = {SECURITY_PANIC_ON, SECURITY_PANIC_OFF}
+
+
+def supported(event: rfxtrxmod.RFXtrxEvent):
+    """Return whether an event supports sirens."""
+    device = event.device
+
+    if isinstance(device, rfxtrxmod.ChimeDevice):
+        return True
+
+    if isinstance(device, rfxtrxmod.SecurityDevice) and isinstance(
+        event, rfxtrxmod.SensorEvent
+    ):
+        if event.values["Sensor Status"] in SECURITY_PANIC_ALL:
+            return True
+
+    return False
+
+
+def get_first_key(data: dict[int, str], entry: str) -> int:
+    """Find a key based on the items value."""
+    return next((key for key, value in data.items() if value == entry))
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up config entry."""
+
+    def _constructor(
+        event: rfxtrxmod.RFXtrxEvent,
+        auto: rfxtrxmod.RFXtrxEvent | None,
+        device_id: DeviceTuple,
+        entity_info: dict,
+    ):
+        """Construct a entity from an event."""
+        device = event.device
+
+        if isinstance(device, rfxtrxmod.ChimeDevice):
+            return [
+                RfxtrxChime(
+                    event.device,
+                    device_id,
+                    entity_info.get(
+                        CONF_SIGNAL_REPETITIONS, DEFAULT_SIGNAL_REPETITIONS
+                    ),
+                    auto,
+                )
+            ]
+
+        if isinstance(device, rfxtrxmod.SecurityDevice) and isinstance(
+            event, rfxtrxmod.SensorEvent
+        ):
+            if event.values["Sensor Status"] in SECURITY_PANIC_ALL:
+                return [
+                    RfxtrxSecurityPanic(
+                        event.device,
+                        device_id,
+                        entity_info.get(
+                            CONF_SIGNAL_REPETITIONS, DEFAULT_SIGNAL_REPETITIONS
+                        ),
+                        auto,
+                    )
+                ]
+
+    await async_setup_platform_entry(
+        hass, config_entry, async_add_entities, supported, _constructor
+    )
+
+
+class RfxtrxTimeoutMixin(Entity):
+    """Mixin to support timeouts on data."""
+
+    _timeout: CALLBACK_TYPE | None = None
+    _timeout_seconds: Final = 2.0
+
+    def _setup_timeout(self):
```

Comment:
We shouldn't set up a timeout unless the siren natively supports that. We can only support device native APIs.

Reply:
We've had this discussion many times. The device have no native indication of off. Their indication of off is absence of command. If you have a good way to support this in the library can you please provide a pull request for that?

Reply:
No. Then I suggest we set `assumed_state` to True and return None from `is_on` to mean unknown state.

This will allow Home Assistant to turn the device on/off and not show any state that we don't know is true.

Reply:
Then it is not usefull to use in automations. For example can't send notifications that the fire alarm in your house is active. Which I think is quite important (my main use case anyway).
The state is known for these devices. Battery might not be know, but we do know if the alarm is active or not.

Reply:
How do we know the state if the device doesn't tell us when it turns the siren off?

Reply:
As mentioned , while siren is on, it sends siren on every X seconds. When it turns off, it stops sending.

This is how linked fire alarms work. While they receive a message that another alarm is ringing, they ring themself. Once they stop getting that, they stop ringing.

Technically we should cyclically send as well if the turn on was requested by us. That is missing here, but we must also support the externally controlled case we we are just displaying state.

Reply:
No, you didn't mention that the siren keeps sending siren on every x seconds as long as it's on. That's a big clarification. Thanks!

In that case this is ok, but then please add a code comment about this.

Do we know that all devices use a two second interval for their messages?

Reply:
Actually no. The ones I've tested with sends it that often (or more often). But we already have a config option for this that it seems I've forgotten to hook up.